src/java.desktop/share/classes/javax/swing/JTable.java

```diff
@@ -2146,42 +2146,56 @@ public boolean getCellSelectionEnabled() {
+    private void selectRows(int rowCount) {
+        ListSelectionModel selModel = selectionModel;
+        selModel.setValueIsAdjusting(true);
+        int oldLead = getAdjustedIndex(selModel.getLeadSelectionIndex(), true);
+        int oldAnchor = getAdjustedIndex(selModel.getAnchorSelectionIndex(), true);
 
-            selModel = selectionModel;
-            selModel.setValueIsAdjusting(true);
-            oldLead = getAdjustedIndex(selModel.getLeadSelectionIndex(), true);
-            oldAnchor = getAdjustedIndex(selModel.getAnchorSelectionIndex(), true);
+        setRowSelectionInterval(0, rowCount - 1);
 
-            setRowSelectionInterval(0, getRowCount()-1);
+        // this is done to restore the anchor and lead
+        SwingUtilities2.setLeadAnchorWithoutSelection(selModel, oldLead, oldAnchor);
 
-            // this is done to restore the anchor and lead
-            SwingUtilities2.setLeadAnchorWithoutSelection(selModel, oldLead, oldAnchor);
+        selModel.setValueIsAdjusting(false);
+    }
 
-            selModel.setValueIsAdjusting(false);
+    private void selectColumns(int columnCount) {
+        ListSelectionModel selModel = columnModel.getSelectionModel();
+        selModel.setValueIsAdjusting(true);
+        int oldLead = getAdjustedIndex(selModel.getLeadSelectionIndex(), false);
+        int oldAnchor = getAdjustedIndex(selModel.getAnchorSelectionIndex(), false);
 
-            selModel = columnModel.getSelectionModel();
-            selModel.setValueIsAdjusting(true);
-            oldLead = getAdjustedIndex(selModel.getLeadSelectionIndex(), false);
-            oldAnchor = getAdjustedIndex(selModel.getAnchorSelectionIndex(), false);
+        setColumnSelectionInterval(0, columnCount - 1);
 
-            setColumnSelectionInterval(0, getColumnCount()-1);
+        // this is done to restore the anchor and lead
+        SwingUtilities2.setLeadAnchorWithoutSelection(selModel, oldLead, oldAnchor);
+
+        selModel.setValueIsAdjusting(false);
+    }
 
-            // this is done to restore the anchor and lead
-            SwingUtilities2.setLeadAnchorWithoutSelection(selModel, oldLead, oldAnchor);
+    /**
+     *  Selects all rows, columns, and cells in the table.
+     */
+    public void selectAll() {
+        int rowCount = getRowCount();
+        int columnCount = getColumnCount();
 
-            selModel.setValueIsAdjusting(false);
+        // If I'm currently editing, then I should stop editing
+        if (isEditing()) {
+            removeEditor();
+        }
+        if (rowCount > 0 && columnCount > 0) {
+            selectRows(rowCount);
+            selectColumns(columnCount);
+        } else if (rowCount > 0 && columnCount == 0) {
+            if (getRowSelectionAllowed()) {
+                selectRows(rowCount);
+            }
+        } else if (columnCount > 0  && rowCount == 0) {
+            if (getColumnSelectionAllowed()) {
+                selectColumns(columnCount);
+            }
```

Comment:
Can we simplify the code from line 2188 - 2198, since the intension is to selectRows/Columns if its available.
`
if (rowCount > 0 ) { 

        if (getRowSelectionAllowed()) { 

                selectRows(rowCount);

            }

        } else if (columnCount > 0 ) {

            if (getColumnSelectionAllowed()) {

                selectColumns(columnCount);

            }`

Reply:
No that might change the existing behavior and JCK expectation for row>0 and column>0 if we introduce the check..
SO, I will like to keep it as it is..

You can raise a followon bug if you would like it to be worked on later..

Reply:
Sure, raising a follow up bug makes sense.